ChristopherChudzicki/threestrap
# Cc/fix tests

### Relevant Issue
#23

### What does this PR do?
This PR fixes the tests (and a few bugs in process) but also overhauls a few other dev-process things:
1. **Uncommits the build directory**
2. **Adds typescript; converts config files to TS** I have found TS extremely helpful for config files, and I needed to make some config file edits, so I went ahead and added TS. I don't anticipate adding any TS source code anytime soon, but I definitely want to add some type declaration files, and I *might* convert some of the test files to typescript.
3. **No more gulp** because...no need for it.
4. Use un-minified Three.js in the example files. _Easier debugging._
5. Adds a watch mode (`npm run dev`) and a test watch mode (`npm run test:watch`)
6. **Fixes tests:**
    - Primarily, the tests were failing because Karma[^1] was trying to run some Node-style ESM code in browsers. It either needs to be browser-suitable ESM, or it needs to be transpiled. I decided to fix this by transpiling, like we do in Mathbox. Using browser-safe ESM might be a nice thing to do in the future, but more than I want tot tackle now.[^2]
    - Some tests were failing because of outdated Jasmine functions. (E.g., `waitFor` and `runs` are no longer supported.) _I updated Jasmine, too, but our old version didn't support these commands. I suspect at some point Jasmine was updated without fixing the tests_
    - Fixed a few bugs, which I'll call out in the review comments.

### How could this be manually tested?

1. `npm install`
2. Try out all the commands in package.json
3. See comments on bugfixes for manual testing notes

[^1]: Karma is an older JS test-runner, still used but not that popular (See [npm trends](https://npmtrends.com/cypress-vs-jest-vs-karma-vs-playwright), "All Time"). Karma runs tests in a real browser, as opposed to something like Jest that runs tests in NodeJS. Running the tests in Karma seems reasonable: Threestrap uses (indirectly, via ThreeJS) some APIs (webgl stuff) that Node/JSDom do not support. That said, very little of Threestrap's code actually uses those Browser-only APIs (possibly none, but I think a small nonzero amount). Mocking ThreeJS and running in Jest seems like it might be reasonable. **But even if we do that someday, I'd rather have the old Karma tests working first.** 

[^2]: Karma does support ES Modules, so in theory we could run our source code in a browser as a module. The biggest hurdle here is bare imports like `import whatever from "three/path/to/thing"`: Browsers do not support such "bare" imports. ("bare" = not absolute or relative or a URL). I believe we would either need to not have any bare imports (and our dependencies couldn't use them either) or we could use an [importmap](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/importmap). Transpiling seems easier.
